server/src/main/java/de/uftos/utils/LessonsDeleter.java

```diff
@@ -0,0 +1,98 @@
+import de.uftos.repositories.database.TimetableRepository;
+import java.util.List;
+
+/**
+ * Delete all lessons that contain all given entities.
+ */
+public class LessonsDeleter {
+
+  private final LessonRepository lessonRepository;
+  private final TimetableRepository timetableRepository;
+
+  /**
+   * Creates the lesson deleter.
+   *
+   * @param lessonRepository    the repository for accessing the lesson table.
+   * @param timetableRepository the repository for accessing the timetable table.
+   */
+  public LessonsDeleter(LessonRepository lessonRepository,
+                        TimetableRepository timetableRepository) {
+    this.lessonRepository = lessonRepository;
+    this.timetableRepository = timetableRepository;
+  }
+
+  /**
+   * Deletes all lessons that use these teachers and all timetables tha
+   * use these lessons.
+   *
+   * @param teachers the teachers for which the lessons have to be deleted.
+   */
+  public void fromTeachers(List<Teacher> teachers) {
+    List<Lesson> lessons = this.lessonRepository.findAllByTeacher(teachers);
+    deleteLessonsAndTimetables(lessons);
+  }
+
+  /**
+   * Deletes all lessons that use these subjects and all timetables tha
+   * use these lessons.
+   *
+   * @param subjects the subjects for which the lessons have to be deleted.
+   */
+  public void fromSubjects(List<Subject> subjects) {
+    List<Lesson> lessons = this.lessonRepository.findAllBySubject(subjects);
+    deleteLessonsAndTimetables(lessons);
+  }
+
+  /**
+   * Deletes all lessons that use these rooms and all timetables tha
+   * use these lessons.
+   *
+   * @param rooms the rooms for which the lessons have to be deleted.
+   */
+  public void fromRooms(List<Room> rooms) {
+    List<Lesson> lessons = this.lessonRepository.findAllByRoom(rooms);
+    deleteLessonsAndTimetables(lessons);
+  }
+
+
+  /**
+   * Deletes all lessons that use these student groups and all timetables tha
+   * use these lessons.
+   *
+   * @param studentGroups the studentGroups for which the lessons have to be deleted.
+   */
+  public void fromStudentGroups(List<StudentGroup> studentGroups) {
+    List<Lesson> lessons = this.lessonRepository.findAllByStudentGroup(studentGroups);
+    deleteLessonsAndTimetables(lessons);
+  }
+
+  /**
+   * Deletes all lessons that use these timeslots and all timetables tha
+   * use these lessons.
+   *
+   * @param timeslots the timeslots for which the lessons have to be deleted.
+   */
+  public void fromTimeSlots(List<Timeslot> timeslots) {
+    List<Lesson> lessons = this.lessonRepository.findAllByTimeslot(timeslots);
+    deleteLessonsAndTimetables(lessons);
+  }
+
+  private void deleteLessonsAndTimetables(List<Lesson> lessons) {
+    List<Timetable> timetables = this.timetableRepository.findAllByLessons(
+        lessons.stream().map(Lesson::getId).toList()
+    );
+    this.lessonRepository.deleteAll(lessons);
+
+    this.timetableRepository.deleteAll(timetables);
```

Comment:
Will this always delete the entire timetable, whenever lessons of the timetable are deleted?
If yes: I think you would need to delete all lessons of the timetable first and not only the once used to find the timetable.
If no: I think the timetables should be filtered so timetables are only deleted if they have no more lessons left.

Reply:
Yes, it deletes the entire timetable. You're right, I need to delete all the lessons from the timetable first, thank you.